File: collector/cc/exe.go
package cc

import (
	"fmt"
	"path/filepath"
	"strings"

	"github.com/ankeesler/btool/collector"
	"github.com/ankeesler/btool/collector/sorter"
	"github.com/ankeesler/btool/log"
	"github.com/ankeesler/btool/node"
	"github.com/pkg/errors"
)

type Exe struct {
	rf ResolverFactory
}

func NewExe(rf ResolverFactory) *Exe {
	return &Exe{
		rf: rf,
	}
}

func (e *Exe) Consume(s collector.Store, n *node.Node) error {
	if filepath.Ext(n.Name) != "" {
		return nil
	}

	c := s.Get(n.Name + ".c")
	cc := s.Get(n.Name + ".cc")
	if c != nil && cc != nil {
		return fmt.Errorf("ambiguous exe %s: %s or %s?", n, c, cc)
	} else if c == nil && cc == nil {
		return fmt.Errorf("no known source for exe %s", n)
	}

	var r node.Resolver
	var err error
	objs := make(map[string]*node.Node)
	if c != nil {
		log.Debugf("exe %s has source %s", n, c)

		err = collectObjs(s, c, ".c", objs)
		r = e.rf.NewLinkC()
	} else { // cc
		log.Debugf("exe %s has source %s", n, cc)

		err = collectObjs(s, cc, ".cc", objs)
		r = e.rf.NewLinkCC()
	}

	if err != nil {
		return errors.Wrap(err, "collect objs")
	}

	libraries := make(map[*node.Node]bool)
	for _, obj := range objs {
		n.Dependency(obj)

		if err := collectLibraries(s, obj, libraries); err != nil {
			return errors.Wrap(err, "collect libraries")
		}
	}
	for library := range libraries {
		n.Dependency(library)
	}
	n.Resolver = r

	sorter.New().Sort(n)
	s.Set(n)

	return nil
}

// The provided n should be a c/cc file!
func collectObjs(
	s collector.Store,
	n *node.Node,
	ext string,
	objs map[string]*node.Node,
) error {
	obj := s.Get(strings.ReplaceAll(n.Name, ext, ".o"))
	if obj == nil {
		return fmt.Errorf("no know object for %s", n)
	}
	if _, ok := objs[obj.Name]; ok {
		return nil
	}
	objs[obj.Name] = obj
	log.Debugf("remembering object %s", obj)

	for _, d := range n.Dependencies {
		if !strings.HasSuffix(d.Name, ".h") {
			continue
		}

		src := s.Get(strings.ReplaceAll(d.Name, ".h", ext))
		if src == nil {
			log.Debugf("skipping header-only dependency %s", d.Name)
			continue
		}

		if err := collectObjs(s, src, ext, objs); err != nil {
			return err
		}
	}

	return nil
}

func collectLibraries(
	s collector.Store,
	n *node.Node,
	libraries map[*node.Node]bool,
) error {
	return node.Visit(n, func(vn *node.Node) error {
		// TODO: this is jank, we should have more of a better interface for this.
		if libs, ok := vn.Labels[LabelLibraries]; ok {
			for _, lib := range strings.Split(libs, ",") {
				if lib == "" {
					continue
				}

				libN := s.Get(lib)
				if libN == nil {
					return fmt.Errorf("unknown node for library %s from node %s", lib, vn)
				}

				libraries[libN] = true
			}
		}
		return nil
	})
}

File: source/app/collector/cc/exe.cc
#include "app/collector/cc/exe.h"

#include <string>
#include <vector>

#include "app/collector/cc/properties.h"
#include "app/collector/properties.h"
#include "app/collector/store.h"
#include "log.h"
#include "node/node.h"
#include "util/fs/fs.h"
#include "util/string/string.h"
#include "util/util.h"

namespace btool::app::collector::cc {

void Exe::OnNotify(::btool::app::collector::Store *s, const std::string &name) {
  if (::btool::util::fs::Ext(name) != "") {
    return;
  }

  auto n = s->Get(name);
  if (n == nullptr) {
    return;
  }

  std::string ext = ".c";
  auto src = s->Get(name + ext);
  if (src == nullptr) {
    ext = ".cc";
    src = s->Get(name + ext);
    if (src == nullptr) {
      AddError("cannot find source for exe " + name);
      return;
    }
  }

  std::vector<::btool::node::Node *> objs;
  if (!CollectObjects(s, src, ext, &objs)) {
    return;
  }
  for (auto obj : objs) {
    n->dependencies()->push_back(obj);
  }

  std::vector<::btool::node::Node *> libs;
  if (!CollectLibraries(s, *n, &libs)) {
    return;
  }
  for (auto lib : libs) {
    n->dependencies()->push_back(lib);
  }

  auto r = (ext == ".cc" ? rf_->NewLinkCC() : rf_->NewLinkC());
  n->set_resolver(r);

  Notify(s, n->name());
}

bool Exe::CollectObjects(::btool::app::collector::Store *s,
                         ::btool::node::Node *n, const std::string &ext,
                         std::vector<::btool::node::Node *> *objs) {
  DEBUGS() << "collect from src " << n->name() << std::endl;

  auto obj_name = ::btool::util::string::Replace(n->name(), ext, ".o");
  auto obj = s->Get(obj_name);
  if (obj == nullptr) {
    AddError("cannot find obj for obj name " + obj_name);
    return false;
  }
  if (::btool::util::Contains(*objs, obj)) {
    return true;
  }
  objs->push_back(obj);

  for (auto d : *n->dependencies()) {
    DEBUGS() << "considering dependency " << d->name() << std::endl;
    if (::btool::util::string::HasSuffix(d->name().c_str(), ".h")) {
      auto src_name = ::btool::util::string::Replace(d->name(), ".h", ext);
      auto src = s->Get(src_name);
      if (src == nullptr) {
        DEBUGS() << "no src for src_name " << src_name << std::endl;
      } else {
        bool success = CollectObjects(s, src, ext, objs);
        if (!success) {
          return false;
        }
      }
    }
  }

  return true;
}

bool Exe::CollectLibraries(::btool::app::collector::Store *s,
                           const ::btool::node::Node &n,
                           std::vector<::btool::node::Node *> *libs) {
  std::vector<std::string> lib_names;
  ::btool::app::collector::CollectStringsProperties(
      n, &lib_names,
      [](const ::btool::node::PropertyStore *ps)
          -> const std::vector<std::string> * {
        return Properties::Libraries(ps);
      });

  for (const auto &lib_name : lib_names) {
    auto lib = s->Get(lib_name);
    if (lib == nullptr) {
      AddError("unknown lib for lib_name " + lib_name);
      return false;
    }

    libs->push_back(lib);
  }

  return true;
}

}  // namespace btool::app::collector::cc
